eta/core/serial.py

```diff
@@ -670,32 +705,325 @@ def from_dict(cls, d):
+            <elements>: an optional list of element uuids. globbing is done when
+                this is not provided. The appropriate name of this keyword
+                argument is determined by the `_ELE_ATTR` member of the
+                DirectoryContainer subclass
+
+        Raises:
+            ContainerError: if there was an error while creating the container
+        '''
+        self._validate()
+
+        if self._ELE_DIR_ATTR not in kwargs:
+            raise ContainerError(
+                "Missing keyword argument %s is required"
+                % (self._ELE_DIR_ATTR))
+
+        setattr(self, self._ELE_DIR_ATTR, kwargs[self._ELE_DIR_ATTR])
+
+        setattr(
+            self,
+            self._ELE_ATTR,
+            kwargs.get(
+                self._ELE_ATTR,
+                [
+                    os.path.splitext(os.path.basename(i))[0]
+                    for i in etau.multiglob(".json",
+                                            root=self._ele_dir+"/**/*")
+                ]
+            )
+        )
+
+    def __iter__(self):
+        return (self._load_ele(path) for path in self._paths)
+
+    def __getitem__(self, idx):
+        return self._load_ele(self._ele_path(idx))
+
+    def __setitem__(self, idx, ele):
+        ele.write_json(self._ele_path(idx))
+
+    def __delitem__(self, idx):
+        etau.delete_file(self._ele_path(idx))
+        super(DirectoryContainer, self).__delitem__(idx)
+
+    def add(self, ele):
+        self.append(ele)
+
+    def add_container(self, container):
+        '''Add a another containers elements to this one. Takes any valid
+        Container subclass (Not just DirectoryContainer subclasses).
+
+        Args:
+            container (Container): the container to add
+        '''
+        for ele in container:
+            self.add(ele)
+
+    def copy(self, new_ele_dir):
+        '''Copy this container deeply.
+
+        Args:
+            new_ele_dir (str): path the store the copy's elements
+
+        Returns:
+            DirectoryContainer
+        '''
+        try:
+            etau.delete_dir(new_ele_dir)
+        except OSError:
+            pass
+        container = self.__class__(**{self._ELE_DIR_ATTR: new_ele_dir})
+        container.add_container(self)
+        return container
+
+    def append(self, ele):
+        '''Append an element.'''
+        self[len(self)] = ele
+
+    def filter_elements(self, filters, match=any):
+        '''Removes elements that don't match the given filters from the
+        container.
+
+        Args:
+            filters: a list of functions that accept elements and return
+                True/False
+            match: a function (usually `any` or `all`) that accepts an iterable
+                and returns True/False. Used to aggregate the outputs of each
+                filter to decide whether a match has occurred. The default is
+                `any`
+        '''
+        self._filter_elements(filters, match)
```

Comment:
I'm confused: it looks like this doesn't actually remove elements, just returns those that match the filter.  Am I missing something?

Reply:
Yes, realizing now that this function as well as `_filter_elements()` needs more work.

Reply:
I've updated the functionality, though I haven't been able to test it yet.